src/pkg/sources/remote.go

```diff
@@ -0,0 +1,190 @@
+}
+
+// LoadPackageMetadata loads a Zarf package's metadata from a remote bundle
+func (r *RemoteBundle) LoadPackageMetadata(dst *layout.PackagePaths, _ bool, _ bool) (err error) {
+	root, err := r.Remote.FetchRoot()
+	if err != nil {
+		return err
+	}
+	var zarfManifest *oci.ZarfOCIManifest
+	for _, layer := range root.Layers {
+		if layer.Digest.Encoded() == r.PkgManifestSHA {
+			zarfManifest, err = r.Remote.FetchManifest(layer)
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+
+	if zarfManifest == nil {
+		return fmt.Errorf(fmt.Sprintf("Zarf image manifest with sha %s not found", r.PkgManifestSHA))
+	}
+
+	for _, layer := range zarfManifest.Layers {
+		if layer.Annotations[ocispec.AnnotationTitle] == config.ZarfYAML {
+			zarfYamlBytes, err := r.Remote.FetchLayer(layer)
+			if err != nil {
+				return err
+			}
+			if err := os.WriteFile(filepath.Join(dst.Base, config.ZarfYAML), zarfYamlBytes, 0644); err != nil {
+				return err
+			}
+			break
+		}
+	}
+
+	return nil
+}
+
+// Collect doesn't need to be implemented
+func (r *RemoteBundle) Collect(_ string) (string, error) {
+	return "", nil
+}
+
+// downloadPkgFromRemoteBundle downloads a Zarf package from a remote bundle
+func (r *RemoteBundle) downloadPkgFromRemoteBundle() error {
+	// todo: use oras.Copy for faster downloads
+	ctx := context.TODO()
+	rootManifest, err := r.Remote.FetchRoot()
+	if err != nil {
+		return err
+	}
+
+	pkgManifestDesc := rootManifest.Locate(r.PkgManifestSHA)
+	if oci.IsEmptyDescriptor(pkgManifestDesc) {
+		return fmt.Errorf("package %s does not exist in this bundle", r.PkgManifestSHA)
+	}
+	// hack to Zarf media type so that FetchManifest works
+	pkgManifestDesc.MediaType = oci.ZarfLayerMediaTypeBlob
+	pkgManifest, err := r.Remote.FetchManifest(pkgManifestDesc)
+	if err != nil || pkgManifest == nil {
+		return err
+	}
+
+	// including the package manifest uses some ORAs FindSuccessors hackery to expand the manifest into all layers
+	// as oras.Copy was designed for resolving layers via a manifest reference, not a manifest embedded inside of another
+	// image
+	layersToPull := []ocispec.Descriptor{pkgManifestDesc}
+	for _, layer := range pkgManifest.Layers {
+		// only fetch layers that exist
+		// since optional-components exists, there will be layers that don't exist
+		// as the package's preserved manifest will contain all layers for all components
+		ok, _ := r.Remote.Repo().Blobs().Exists(ctx, layer)
+		if ok {
+			layersToPull = append(layersToPull, layer)
+		}
+	}
+
+	store, err := file.New(r.TmpDir)
+	if err != nil {
+		return err
+	}
+	defer store.Close()
+
+	spinner := message.NewProgressSpinner("Pulling bundled Zarf package")
+	defer spinner.Stop()
+	for _, layer := range layersToPull {
+		spinner.Updatef(fmt.Sprintf("Pulling bundle layer: %s", layer.Digest.Encoded()))
+		lb, err := r.Remote.Repo().Fetch(ctx, layer)
+		if err != nil {
+			return err
+		}
+
+		err = store.Push(ctx, layer, lb)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
this has no concurrency and should use `oras.Copy` if methods in `oci.NewOrasRemote` are insufficient

Reply:
Tracking, we restored OCI concurrency in push/pull ops but currently have a ticket to address it everywhere else: https://github.com/defenseunicorns/uds-cli/issues/115